### PR TITLE
Add more options in choose_qparams_affine for tinygemm op

### PR DESCRIPTION
**Pull request by jerryzh168**
Summary:
This is in preparation for replacing tinygemm q/dq ops with the unified quant primitive ops

tinygemm choose_qparams op (and also quantize/dequantize op): https://github.com/pytorch/ao/blob/main/torchao/quantization/quant_primitives.py#L36 is different from the other choose_qparams op in that it does not enforce zero_point to be exactly representable, meaning the floating point value 0 can't be exactly represented by an integer value in quantized tensor. This PR adds a flag to produce a zero_point value that can be adapted to be used by tinygemm kernels.

Test Plan:
python test/quantization/test_quant_primitives.py -k test_tinygemm_get_groupwise_affine_qparams

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

**Review by HDCharles — DISMISSED** [review]
we need to compare with the existing numerics of choosing qparams and q/dq. If we're introducing error for no benefit then i don't thing we should deprecate the other primitives. At the very least we need to discuss the functional changes first.

also need to test whether these primitives also work with gptq, there are nuances that i'm not sure are being captured.

**Comment by jerryzh168** [comment]
> we need to compare with the existing numerics of choosing qparams and q/dq. If we're introducing error for no benefit then i don't thing we should deprecate the other primitives. At the very least we need to discuss the functional changes first.
> 
> also need to test whether these primitives also work with gptq, there are nuances that i'm not sure are being captured.

sounds good, updated the op to work with tinygemm numerics.

for gptq and other types of e2e quant tests, we could establish a dashboard for perf and accuracy first, we can prioritize this in 0.3

File: test/quantization/test_quant_primitives.py [hunk 1]
```diff
@@ -56,8 +57,8 @@ def test_get_group_qparams_symmetric(self):
         scale_obs = scale_obs.reshape(weight.shape[0], -1)
 
         # assert that scales are identical
-        (scale_ao, _) = get_group_qparams_symmetric(weight, n_bit, groupsize)
-        torch.testing.assert_allclose(scale_obs, scale_ao, rtol=0, atol=0)
+        (scale_ao, _) = get_group_qparams_symmetric(weight, n_bit, groupsize, precision=torch.float16)
+        torch.testing.assert_close(scale_obs, scale_ao, rtol=0, atol=0)
```
Comment (cpuhrsch) [inline test/quantization/test_quant_primitives.py:61]: why this change?

File: torchao/quantization/quant_primitives.py [hunk 2]
```diff
@@ -256,19 +256,24 @@ def choose_qparams_affine(
...
-                           e.g. when size is the same as the input tensor dimension, we are using per tensor quantization
+        block_size: (Tuple[int, ...]): granularity of quantization, this means the size of the tensor elements that's sharing the same qparam
+          e.g. when size is the same as the input tensor dimension, we are using per tensor quantization
         target_dtype (torch.dtype): dtype for target quantized Tensor
         quant_min (Optional[int]): minimum quantized value for target quantized Tensor
         quant_max (Optioanl[int]): maximum quantized value for target quantized Tensor
         eps (Optional[float]): minimum scale, if not provided, default to eps of input.dtype
         scale_dtype (torch.dtype): dtype for scale Tensor
         zero_point_dtype (torch.dtype): dtype for zero_point Tensor
+        _is_zero_exactly_representable (bool): a private flag to indicate whether we need zero to be exactly
```
Comment (cpuhrsch) [inline torchao/quantization/quant_primitives.py:273]: Interesting. I'd expect symmetric without zero_point to imply this. Is that true?

File: test/quantization/test_quant_primitives.py (hunk 1 above)
Comment (jerryzh168) [reply test/quantization/test_quant_primitives.py:61]: you mean assert_close? I saw assert_allclose is deprecated, that's why I updated them

File: torchao/quantization/quant_primitives.py [hunk 3]
```diff
@@ -256,19 +256,24 @@ def choose_qparams_affine(
...
-                           e.g. when size is the same as the input tensor dimension, we are using per tensor quantization
+        block_size: (Tuple[int, ...]): granularity of quantization, this means the size of the tensor elements that's sharing the same qparam
+          e.g. when size is the same as the input tensor dimension, we are using per tensor quantization
         target_dtype (torch.dtype): dtype for target quantized Tensor
         quant_min (Optional[int]): minimum quantized value for target quantized Tensor
         quant_max (Optioanl[int]): maximum quantized value for target quantized Tensor
         eps (Optional[float]): minimum scale, if not provided, default to eps of input.dtype
         scale_dtype (torch.dtype): dtype for scale Tensor
         zero_point_dtype (torch.dtype): dtype for zero_point Tensor
+        is_exact_zero (bool): a flag to indicate whether we need zero to be exactly
```
Comment (cpuhrsch) [inline torchao/quantization/quant_primitives.py:273]: nit: Can you also define what "zero being exactly representable" means for the outputs? You can add `zero_padding` as an example of when this is useful`. Also I'd remove the `is_` part and choose `preserve_zero` or some other verb. Since this is a kwarg to a function and not a property of a class it seems more consistent.

**Review by cpuhrsch — APPROVED** [review]

File: torchao/quantization/quant_primitives.py [hunk 4]
```diff
@@ -288,17 +298,26 @@ def choose_qparams_affine(
...
+        min_val_neg = torch.min(min_val, torch.zeros_like(min_val))
+        max_val_pos = torch.max(max_val, torch.zeros_like(max_val))
+    else:
+        min_val_neg = min_val
+        max_val_pos = max_val
 
     if mapping_type == MappingType.SYMMETRIC:
         max_val_pos = torch.max(-min_val_neg, max_val_pos)
         scale = max_val_pos / (float(quant_max - quant_min) / 2)
+        assert preserve_zero, "non-representable zero path is not implemented for symmetric quantization"
```
Comment (cpuhrsch) [inline torchao/quantization/quant_primitives.py:311]: Adding a `assertRaisesRegex` based test for this as a follow up would be good. Also you can raise a ValueError. This prevents us from accidentally adding coverage later on even though it might mathematically not make sense.

Comment (jerryzh168) [reply torchao/quantization/quant_primitives.py:311]: OK sounds good
